frozen array in XRInputSourcesChangeEvent, update `to_frozen_array` doc

chickenleaf opened the pull request:
PR Description
Issue:

1. `XRInputSourcesChangeEvent` was not using `to_frozen_array` for array conversion, leading to manual conversions inconsistent with WebXR spec requirements. 
2. `to_frozen_array` rustdoc comment needs to be updated.

Changes:
1. Replaced manual conversions with `to_frozen_array`
2. Used `JSContext::from_ptr(*cx)` for context creation
3. Changed `to_frozen_array` rustdoc comment to reflect broader usage
4. Add tests to ensure `XRInputSourcesChangeEvent` `.added` and `.removed` are frozen arrays, with assertions that modifications throw `TypeError`.

Test:
`tests/wpt/tests/webxr/events_input_sources_change.https.html`
---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #34050 

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

chickenleaf committed:
frozen array in XRInputSourcesChangeEvent, update 	o_frozen_array doc

Signed-off-by: L Ashwin B <lashwinib@gmail.com>

jdm commented on components/script/dom/xrinputsourceschangeevent.rs:
```diff
@@ -88,12 +88,12 @@ impl XRInputSourcesChangeEvent {
-            changeevent.removed.set(removed_val.get());
+            changeevent
+                .added
+                .set(to_frozen_array(added, JSContext::from_ptr(*cx)));
```
Oh, we can just pass `cx` here instead of dereferencing it and recreating the JSContext type from the pointer.

chickenleaf replied:
done!
also shall i remove this?
```
warning: unnecessary `unsafe` block
  --> components\script\dom\xrinputsourceschangeevent.rs:90:9
   |
90 |         unsafe {
   |         ^^^^^^ unnecessary `unsafe` block
   |
   = note: `#[warn(unused_unsafe)]` on by default
```

jdm replied:
Yep!

jdm reviewed (approved):
This looks fine, but we can simplify it a little bit!

jdm commented:
>These changes do not require tests because they do not affect the functionality.

This isn't true; we're definitely changing functionality here. We could update the webxr tests to verify that the added and removed properties are frozen arrays. There's similar test code that looks like this: https://github.com/web-platform-tests/wpt/blob/e7e2de2d3ad8e8160509838ddbf37666cd69a90e/push-api/supported-encodings.https.any.js#L16

chickenleaf commented:
> >These changes do not require tests because they do not affect the functionality.
> 
> This isn't true; we're definitely changing functionality here. We could update the webxr tests to verify that the added and removed properties are frozen arrays. There's similar test code that looks like this: https://github.com/web-platform-tests/wpt/blob/e7e2de2d3ad8e8160509838ddbf37666cd69a90e/push-api/supported-encodings.https.any.js#L16

Which tests need the updation?

chickenleaf committed:
simplified changes

Signed-off-by: L Ashwin B <lashwinib@gmail.com>

chickenleaf committed:
Merge remote-tracking branch 'origin/main' into input-array-conversion

jdm commented:
https://github.com/web-platform-tests/wpt/blob/e7e2de2d3ad8e8160509838ddbf37666cd69a90e/webxr/events_input_sources_change.https.html looks like a good candidate for adding the new test coverage.

chickenleaf commented:
> https://github.com/web-platform-tests/wpt/blob/e7e2de2d3ad8e8160509838ddbf37666cd69a90e/webxr/events_input_sources_change.https.html looks like a good candidate for adding the new test coverage.

shall i update it to do this as well? 

> tests to verify that the added and removed properties are frozen arrays

just adding two tests to see if they throw a typeerror when i try to modify the arrays.
similar to the `push-api/supported-encodings.https.any.js` L16 u sent

jdm commented:
Yep! That would be great!

chickenleaf commented:
![Screenshot 2024-11-02 165259](https://github.com/user-attachments/assets/85a7e889-c4f0-483a-8e1a-974e9c46fc23)
this is good?
```
assert_throws_js(
    TypeError,
    () => { event.added[0] = "test"; },
    "added array must be frozen"
);
assert_throws_js(
    TypeError,
    () => { event.removed[0] = "test"; },
    "removed array must be frozen"
);
```
and added use strict on top. should i be adding any comments? 

also,,, is using `array.push()` better? compared to this way?

chickenleaf committed:
Merge remote-tracking branch 'origin/main' into input-array-conversion

jdm commented:
That looks reasonable! Does the test pass?

chickenleaf committed:
added tests

Signed-off-by: L Ashwin B <lashwinib@gmail.com>

chickenleaf commented:
> That looks reasonable! Does the test pass?

weirdly enough no? the whole test doesnt pass, but the subtests pass

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#48944) with upstreamable changes.
